crates/turbopack-trace-server/src/self_time_tree.rs

```diff
@@ -0,0 +1,248 @@
+                            left: take(left_right),
+                            split_point: *split_point,
+                            right: take(right),
+                        })),
+                    };
+                    *split_point = *left_split_point;
+                    *left = take(left_left);
+                    let entries = replace(&mut self.entries, left_entries);
+                    self.count = left.count + right.count + self.entries.len();
+                    for SelfTimeEntry { start, end, item } in entries {
+                        self.insert(start, end, item);
+                    }
+                }
+            } else if *right_count > *left_count * 3 + right_entries.len() {
+                if let Some(box SelfTimeChildren {
+                    left: right_left,
+                    split_point: right_split_point,
+                    right: right_right,
+                }) = right_children
+                {
+                    let right_entries = take(right_entries);
+                    *left = Self {
+                        count: left.count + right_left.count + self.entries.len(),
+                        entries: take(&mut self.entries),
+                        children: Some(Box::new(SelfTimeChildren {
+                            left: take(left),
+                            split_point: *split_point,
+                            right: take(right_left),
+                        })),
+                    };
+                    *split_point = *right_split_point;
+                    *right = take(right_right);
+                    let entries = replace(&mut self.entries, right_entries);
+                    self.count = left.count + right.count + self.entries.len();
+                    for SelfTimeEntry { start, end, item } in entries {
+                        self.insert(start, end, item);
+                    }
+                }
+            }
+        }
+    }
+
+    pub fn lookup_range_count(&self, start: u64, end: u64) -> u64 {
+        let mut total_count = 0;
+        for entry in &self.entries {
+            if entry.start < end && entry.end > start {
+                let start = max(entry.start, start);
+                let end = min(entry.end, end);
+                let span = end - start;
+                total_count += span;
+            }
+        }
+        if let Some(children) = &self.children {
+            if start < children.split_point {
+                total_count += children.left.lookup_range_count(start, end);
+            }
+            if end > children.split_point {
+                total_count += children.right.lookup_range_count(start, end);
+            }
+        }
+        total_count
+    }
+
+    pub fn for_each_in_range(&self, start: u64, end: u64, mut f: impl FnMut(u64, u64, &T)) {
+        self.for_each_in_range_ref(start, end, &mut f);
+    }
+
+    fn for_each_in_range_ref(&self, start: u64, end: u64, f: &mut impl FnMut(u64, u64, &T)) {
```

Comment:
I think `&mut dyn` would be better as this is a sort of `#[cold]` path for normal runs, but well it's a super minor issue